bitcoin/examples/sighash.rs

```diff
@@ -0,0 +1,238 @@
+                        script_pubkey_bytes = pb.as_bytes();
+                    } else {
+                        //All others from 1 (0-indexed) to M-2 must be signatures
+                        let sig_bytes = pb.as_bytes();
+                        let sig = bitcoin::ecdsa::Signature::from_slice(sig_bytes)
+                            .expect("failed to parse sig");
+                        let sig_len = sig_bytes.len() - 1; //last byte is EcdsaSighashType sighash flag
+                                                           //ECDSA signature in DER format lengths are between 70 and 72 bytes
+                        assert!(
+                            (70..=72).contains(&sig_len),
+                            "signature length {} out of bounds",
+                            sig_len
+                        );
+                        //take sighash_flag into account - can they be different?
+                        assert!(
+                            last_sighash_flag == bitcoin::sighash::EcdsaSighashType::None
+                                || last_sighash_flag == sig.hash_ty,
+                            "different sighash flags"
+                        );
+                        last_sighash_flag = sig.hash_ty;
+                    }
+                    println!("ScriptSig PushBytes len {}: {:?}", pb.as_bytes().to_vec().len(), pb)
+                }
+                Instruction::Op(op) => {
+                    unreachable!("we only expect PushBytes here, got Op({})", op)
+                }
+            }
+        }
+    }
+
+    let script_code = bitcoin::Script::from_bytes(script_pubkey_bytes);
+    let cache = sighash::SighashCache::new(&tx);
+    let sighash = cache
+        .legacy_signature_hash(inp_idx, script_code, last_sighash_flag.to_u32())
+        .expect("failed to compute sighash");
+    println!("Legacy sighash: {:x}\n", sighash);
+}
+
+/// Computes sighash for a segwit multisig transaction input that spends a p2wsh output with "witness_v0_scripthash" scriptPubKey.type
+///
+/// # Arguments
+///
+/// * `raw_tx` - spending tx hex
+/// * `inp_idx` - spending tx input index
+/// * `value` - ref tx output value in sats
+fn compute_sighash_p2wsh(mut raw_tx: &[u8], inp_idx: usize, value: u64) {
+    let tx: bitcoin::Transaction =
+        bitcoin::consensus::Decodable::consensus_decode(&mut raw_tx).unwrap();
+    let inp = &tx.input[inp_idx];
+    let witness = &inp.witness;
+    println!("witness {:?}", witness);
+    let mut last_sighash_flag = bitcoin::sighash::EcdsaSighashType::None;
+
+    //last element is called  witnessScript according to BIP141. It supersedes scriptPubKey.
+    let witness_script_bytes: &[u8] = witness.last().expect("Out of Bounds");
+    let witness_script = bitcoin::Script::from_bytes(witness_script_bytes);
+
+    //in an M of N multisig, the witness elements from 1 (0-based) to M-2 are signatures (with sighash flags as the last byte)
+    for n in 1..witness.len() - 2 {
```

Comment:
There is an off by one error here because witness length is 4 so this only loops once (assuming this function is called for a 2-by-2 multisig.